app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/recordaudio/RecordAudioViewModel.kt

```diff
@@ -0,0 +1,323 @@
+    fun onDismissMaxFileSizeReachedDialog()
+    fun discardRecording(onCloseRecordAudio: () -> Unit)
+    fun sendRecording(onAudioRecorded: (UriAsset) -> Unit, onComplete: () -> Unit)
+    fun onPlayAudio()
+    fun onSliderPositionChange(position: Int)
+}
+
+@Suppress("TooManyFunctions")
+@HiltViewModel
+class RecordAudioViewModelImpl @Inject constructor(
+    private val kaliumFileSystem: KaliumFileSystem,
+    private val recordAudioMessagePlayer: RecordAudioMessagePlayer,
+    private val getAssetSizeLimit: GetAssetSizeLimitUseCase,
+    private val observeEstablishedCalls: ObserveEstablishedCallsUseCase,
+    private val currentScreenManager: CurrentScreenManager
+) : RecordAudioViewModel, ViewModel() {
+
+    private var state: RecordAudioState by mutableStateOf(RecordAudioState())
+
+    private var assetLimitInMegabyte by Delegates.notNull<Long>()
+
+    private var mediaRecorder: MediaRecorder? = null
+
+    private var hasOngoingCall: Boolean = false
+
+    private val infoMessage = MutableSharedFlow<UIText>()
+
+    override fun getButtonState(): RecordAudioButtonState = state.buttonState
+
+    override fun getDiscardDialogState(): RecordAudioDialogState = state.discardDialogState
+
+    override fun getPermissionsDeniedDialogState(): RecordAudioDialogState =
+        state.permissionsDeniedDialogState
+
+    override fun getMaxFileSizeReachedDialogState(): RecordAudioDialogState =
+        state.maxFileSizeReachedDialogState
+
+    override fun getOutputFile(): File? = state.outputFile
+
+    override fun getAudioState(): AudioState = state.audioState
+
+    override fun getInfoMessage(): SharedFlow<UIText> = infoMessage.asSharedFlow()
+
+    init {
+        observeAudioPlayerState()
+
+        viewModelScope.launch {
+            assetLimitInMegabyte = getAssetSizeLimit(isImage = false)
+
+            launch {
+                observeAudioFileSize()
+            }
+            launch {
+                observeScreenState()
+            }
+            launch {
+                observeUserIsInCall()
+            }
+        }
+    }
+
+    private suspend fun observeUserIsInCall() {
+        observeEstablishedCalls().collect {
+            hasOngoingCall = it.isNotEmpty()
+        }
+    }
+
+    private suspend fun observeScreenState() {
+        currentScreenManager.observeCurrentScreen(viewModelScope).collect { currentScreen ->
+            if (state.buttonState == RecordAudioButtonState.RECORDING &&
+                currentScreen == CurrentScreen.InBackground
+            ) {
+                stopRecording()
+                discardRecording(onCloseRecordAudio = {})
+            }
+        }
+    }
+
+    private fun observeAudioPlayerState() {
+        viewModelScope.launch {
+            recordAudioMessagePlayer.audioMessageStateFlow.collect {
+                state = state.copy(
+                    audioState = it
+                )
+            }
+        }
+    }
+
+    override fun startRecording(context: Context) {
```

Comment:
We should not use context in viewModel, this can create memory leaks

> Caution: A [ViewModel](https://developer.android.com/reference/androidx/lifecycle/ViewModel) usually shouldn't reference a view, [Lifecycle](https://developer.android.com/reference/androidx/lifecycle/Lifecycle), or any class that may hold a reference to the activity context. Because the ViewModel lifecycle is larger than the UI's, holding a lifecycle-related API in the ViewModel could cause memory leaks.

You may need to make the setup in a different class and inject there the context